cookbooks/cdo-repository/recipes/default.rb

```diff
@@ -3,44 +3,35 @@
 # Recipe:: default
 #
 
-include_recipe "cdo-github-access"
-
-template "/home/#{node[:current_user]}/.gemrc" do
-  source 'gemrc.erb'
-  user node[:current_user]
-  group node[:current_user]
+# Sync repo via SSH if key is provided.
+include_recipe 'cdo-github-access'
+has_ssh_key = node['cdo-github-access'] && node['cdo-github-access']['id_rsa'] != ''
+if has_ssh_key
+  node.override['cdo-repository']['url'] = 'git@github.com:code-dot-org/code-dot-org.git'
 end
 
-git "/home/#{node[:current_user]}/#{node.chef_environment}" do
-  # Clone repo via SSH if key is provided, anonymous-HTTPS otherwise.
-  if node['cdo-github-access'] && node['cdo-github-access']['id_rsa'] != ''
-    repository 'git@github.com:code-dot-org/code-dot-org.git'
-  else
-    repository 'https://github.com/code-dot-org/code-dot-org.git'
-  end
+home_path = node[:home]
+git_path = File.join home_path, node.chef_environment
 
-  # Sync to the production or staging branch as appropriate.
-  branch = (node.chef_environment == 'adhoc') ? 'staging' : node.chef_environment
-  revision branch
+git git_path do
+  repository node['cdo-repository']['url']
+  depth node['cdo-repository']['depth'] if node['cdo-repository']['depth']
 
-  # It is not necessary to checkout the staging branch because we get it automatically
-  # when cloning the repository.
-  enable_checkout (branch != 'staging')
+  # Checkout during clone using --branch [checkout_branch] to skip additional checkout step.
+  enable_checkout false
 
-  # Set the name of the deploy branch to match the environment.
+  branch = node['cdo-repository']['branch']
   checkout_branch branch
+  revision branch
 
-  # Sync the deploy branch to the environment branch.
-  action :checkout
-  user node[:current_user]
-
-  # Set the upstream branch to the appropriate origin.
-  notifies :run, "execute[select-upstream-branch]", :immediately
-end
+  # Default checkout-only for CI-managed instances. (CI script manages pull on updates)
+  action = :checkout
```

Comment:
This code is tricky because of the mutable state in `action`, which can be repeatedly updated by each one of the postfix if statements.  I believe the following variation which gets rid of the mutability is more obvious (also, it avoids having a local and a method with the same name.)
````
# Skip git-repo sync when using a shared volume to prevent data loss on the host.
if GitHelper.shared_volume?(git_path, home_path)
  action :nothing

# Sync instead of checkout for adhoc instances that aren't CI-managed.
elsif node.chef_environment == 'adhoc'
   action :sync

# Default checkout-only for CI-managed instances. (CI script manages pull on updates)
else
  action :checkout
end
````